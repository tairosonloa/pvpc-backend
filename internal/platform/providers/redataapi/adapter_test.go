package redataapi

import (
	"context"
	"net/http"
	"net/http/httptest"
	"os"
	"testing"
	"time"

	"pvpc-backend/internal/domain"

	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/stretchr/testify/require"
)

func Test_FetchPVPCPrices_Success(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		require.Equal(t, http.MethodGet, r.Method)
		require.Equal(t, pvpcPricesEndpoint, r.URL.Path)
		require.Equal(t, "application/json", r.Header.Get("Accept"))
		require.Equal(t, "end_date=2023-09-08T23%3A59&geo_ids=1234&start_date=2023-09-08T00%3A00&time_trunc=hour", r.URL.RawQuery)

		res, err := os.ReadFile("./mocks/fetch_pvpc_response.json")
		require.NoError(t, err)

		w.WriteHeader(http.StatusOK)
		w.Write(res)
	}))
	defer server.Close()

	zone, err := domain.NewZone(domain.ZoneDto{ID: "ZON", ExternalID: "1234", Name: "Zone Name"})
	require.NoError(t, err)

	date, err := time.Parse("2006-01-02T15:04:05Z", "2023-09-08T17:54:36Z")
	require.NoError(t, err)

	adapter := NewREDataAPI(server.URL)
	prices, err := adapter.FetchPVPCPrices(context.Background(), []domain.Zone{zone}, date)
	require.NoError(t, err)
	require.Len(t, prices, 1)
	snaps.MatchSnapshot(t, prices[0].Serialize())
}

func Test_FetchPVPCPrices_Error(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		require.Equal(t, http.MethodGet, r.Method)
		require.Equal(t, pvpcPricesEndpoint, r.URL.Path)
		require.Equal(t, "application/json", r.Header.Get("Accept"))
		require.Equal(t, "end_date=2023-09-08T23%3A59&geo_ids=1234&start_date=2023-09-08T00%3A00&time_trunc=hour", r.URL.RawQuery)

		w.WriteHeader(http.StatusTooManyRequests)
		w.Write([]byte("error"))
	}))
	defer server.Close()

	zone, err := domain.NewZone(domain.ZoneDto{ID: "ZON", ExternalID: "1234", Name: "Zone Name"})
	require.NoError(t, err)

	date, err := time.Parse("2006-01-02T15:04:05Z", "2023-09-08T17:54:36Z")
	require.NoError(t, err)

	adapter := NewREDataAPI(server.URL)
	prices, err := adapter.FetchPVPCPrices(context.Background(), []domain.Zone{zone}, date)
	require.NoError(t, err)
	require.Len(t, prices, 0)
}
