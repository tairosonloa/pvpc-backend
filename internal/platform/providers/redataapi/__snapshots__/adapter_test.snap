
[Test_FetchPVPCPrices_Success - 1]
domain.PricesDto{
    ID:     "ZON-2023-09-08",
    Date:   "2023-09-08",
    Zone:   domain.ZoneDto{ID:"ZON", ExternalID:"1234", Name:"Zone Name"},
    Values: {
        {Datetime:"2023-09-08T00:00:00+02:00", Value:150.95},
        {Datetime:"2023-09-08T01:00:00+02:00", Value:138.86},
        {Datetime:"2023-09-08T02:00:00+02:00", Value:129.29},
        {Datetime:"2023-09-08T03:00:00+02:00", Value:128.75},
        {Datetime:"2023-09-08T04:00:00+02:00", Value:125.1},
        {Datetime:"2023-09-08T05:00:00+02:00", Value:120.4},
        {Datetime:"2023-09-08T06:00:00+02:00", Value:134.12},
        {Datetime:"2023-09-08T07:00:00+02:00", Value:154.64},
        {Datetime:"2023-09-08T08:00:00+02:00", Value:182.07},
        {Datetime:"2023-09-08T09:00:00+02:00", Value:161.71},
        {Datetime:"2023-09-08T10:00:00+02:00", Value:190.61},
        {Datetime:"2023-09-08T11:00:00+02:00", Value:183.07},
        {Datetime:"2023-09-08T12:00:00+02:00", Value:181.32},
        {Datetime:"2023-09-08T13:00:00+02:00", Value:180.03},
        {Datetime:"2023-09-08T14:00:00+02:00", Value:133.42},
        {Datetime:"2023-09-08T15:00:00+02:00", Value:134.11},
        {Datetime:"2023-09-08T16:00:00+02:00", Value:145.75},
        {Datetime:"2023-09-08T17:00:00+02:00", Value:166.61},
        {Datetime:"2023-09-08T18:00:00+02:00", Value:216.17},
        {Datetime:"2023-09-08T19:00:00+02:00", Value:230.7},
        {Datetime:"2023-09-08T20:00:00+02:00", Value:260.96},
        {Datetime:"2023-09-08T21:00:00+02:00", Value:241.9},
        {Datetime:"2023-09-08T22:00:00+02:00", Value:177.06},
        {Datetime:"2023-09-08T23:00:00+02:00", Value:174.08},
    },
}
---
